### PR TITLE
New Gamemode: Name pending (pre-distress / survivor)

### PR DESCRIPTION
**Pull request by psykzz**
## About The Pull Request

Adds a new gamemode that pits unequiped survivors against xeno's. Survivors are required to search the land for items and eventually build a distress beacon and call for help.

```
Survivor gamemode

This is pre-distress beacon. The survivors spawn with different equipment and need to scavenge to find weapons and a beacon. Once they have the beacon setup they need to defend it until the timer drops to 0. At which point ERT spawns and the survivors win. In the meantime Xenos need to do everything they can to stop it.
Survivors have pinpointers to make it easier to find the key items, and spawn with some random thigs to help survival a little.

```

## Why It's Good For The Game

New gamemodes!

## Changelog
:cl:
add: A new gamemode <name pending>
/:cl:

## Todo

- [x] Late join
- [ ] ERT should arrive on the planet.
- [x] Latejoin xeno

### CODE REVIEW

**Comment by Shadowflame909** [comment]
Xeno Mode?

I know just a man who would complain about this!

But he doesn't seem to be here today

File: code/game/gamemodes/survivor.dm [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+SPAWNS
+*/
+
+/obj/effect/landmark/survivor/spawn/human/Initialize()
+    . = ..()
+    GLOB.survivor_spawn_human += loc
+    flags_atom |= INITIALIZED
+    return INITIALIZE_HINT_QDEL
+
+/obj/effect/landmark/survivor/spawn/xeno/Initialize()
```
Comment (Jalleo) [inline code/game/gamemodes/survivor.dm]: Cant you just use the original xeno spawn landmarks?

Comment (psykzz) [reply code/game/gamemodes/survivor.dm]: I could but for now, i just wanted to have my own, this is also the reason i seperated the map out in a seperate commit because i want to redo that last.

File: code/game/gamemodes/extended.dm [hunk 2]
```diff
@@ -4,17 +4,15 @@
 	required_players = 0
 
 
-/datum/game_mode/announce()
+/datum/game_mode/extended/announce()
```
Comment (Jalleo) [inline code/game/gamemodes/extended.dm]: Did you make sure that there was a game_mode/announce() declared elsewhere to ensure there is a debug message encase people mess up with their new mode?

**Comment by psykzz** [comment]
## Playtest 2
### Bugs
- [ ] Ready button broken
- [ ] Beacon can be spit acid on

### Feedback
- [ ] Forced Ovi queen
- [ ] Remove Survivor skills 
- [ ] better weapons (maybe shotgun)
- [ ] Limit late join to 5 mins
- [ ] More admin logs
- [ ] Choose survivor type
- [ ] Limit T3 xeno

**Comment by psykzz** [comment]
So far however, its been 
Survivors: 1
Xenos: 2

The biggest impact seems to be the map.

**Comment by psykzz** [comment]
After a bunch of playtesting, im happy for review. Balance is still an issue, but looking for more feedback now.

File: code/__DEFINES/gamemode.dm [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#define GAMEMODE_SURIVOR_DRAW "Draw"
```
Comment (DominikPanic) [inline code/__DEFINES/gamemode.dm]: There's already mode.dm

File: code/datums/emergency_calls/emergency_call.dm [hunk 4]
```diff
@@ -100,7 +101,7 @@
 		to_chat(usr, "<span class='danger'>You are jobbanned from the emergency reponse team!</span>")
 		return
 
-	if(!istype(distress) || !SSticker.mode.waiting_for_candidates || distress.mob_max < 1)
+	if(!SSticker.mode.waiting_for_candidates || distress.mob_max < 1)
```
Comment (DominikPanic) [inline code/datums/emergency_calls/emergency_call.dm]: Why this change if I may ask?

File: code/game/objects/items/gamemode/survivor.dm [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+    if(!ishuman(user)) 
+        return
+
+    if(!user.mind)
+        to_chat(user, "<span class='warning'>It doesn't seem to do anything for you.</span>")
+        return
+
+    var/area/A = get_area(user)
+    if(A && istype(A) && A.ceiling >= CEILING_METAL)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
    if(istype(A) && A.ceiling >= CEILING_METAL)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        to_chat(user, "<span class='warning'>You have to be outside or under a glass ceiling to activate this.</span>")
+        return
+
+    user.visible_message("<span class='notice'>[user] starts setting up [src] on the ground.</span>",
+    "<span class='notice'>You start setting up [src] on the ground and inputting all the data it needs.</span>")
+    if(do_after(user, activation_time, TRUE, 5, BUSY_ICON_FRIENDLY))
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Early return

File: code/game/objects/items/gamemode/survivor.dm [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+    SetLuminosity(5)
+
+    if (issurvivorgamemode(SSticker.mode))
+        var/datum/game_mode/survivor/GM = SSticker.mode
+        GM.beacon = src
+
+    noise_timer_id = addtimer(CALLBACK(src, .proc/make_noise), 7 SECONDS, TIMER_LOOP|TIMER_STOPPABLE)
+    beacon_timer_id = addtimer(CALLBACK(src, .proc/call_distress_team), distress_timer, TIMER_UNIQUE|TIMER_STOPPABLE)
+
+    for (var/mob/M in GLOB.alive_human_list)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Optimize this with
```
for var/i
var/mob/M = i
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+    else if(iswelder(W) && current_hp == max_hp)
+        to_chat(user, "That doesn't look damaged!")
+
+    // Required Nearby components
+    // TODO: Check if this is wasteful.
+    var/list/nearby = range(2, src)
+    var/nearby_setup = 0
+    for (var/R in required_nearby)
+        var/obj/found = locate(R) in nearby
+        if (found && found.anchored == 1)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
        if (found?.anchored)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+            nearby_setup++
+
+    if(length(required_components) == 0 && current_hp == max_hp && nearby_setup == length(required_nearby))
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
    if(!length(required_components) && current_hp == max_hp && nearby_setup == length(required_nearby))
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+                return
+            if (user.transferItemToLoc(W, src))
+                required_components -= R
+                internal_components += W
+                user.visible_message("<span class='notice'>[user] installed \the [C]</span>","<span class='notice'>You installed \the [C]</span>")
+                break
+
+    // repair dmg
+    if(iswelder(W) && current_hp < max_hp)
+        if(do_after(user, 1 SECONDS, TRUE, 5, BUSY_ICON_BUILD,, TRUE))
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Early return this whole branch

File: code/game/objects/items/gamemode/survivor.dm [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+    var/scan_range = 2
+    var/required_nearby = list(
+        /obj/item/laptop/rescue,
+    )
+
+    // Contents list, for when it gets broken. We spit everything out again
+    var/internal_components = list()
+
+    // Custom health, It can't be destroyed but it can be stopped by reducing health to 0
+    var/max_hp = 100
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Use obj_integritiy for this

File: code/game/objects/items/gamemode/survivor.dm [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+    name = "Communications laptop"
+    desc = "A device used for augmenting communication. Someone seems to have left it logged in with open communications to a nearby ship. Set this up next to a rescue beacon."
+    icon = 'icons/obj/machines/laptop_beacon.dmi'
+    icon_state = "tandy0"
+    luminosity = 4
+    resistance_flags = UNACIDABLE|INDESTRUCTIBLE
+    w_class = WEIGHT_CLASS_BULKY
+    var/activation_time = 5 SECONDS
+
+/obj/item/laptop/rescue/Destroy()
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: This does nothing

File: code/game/objects/items/gamemode/survivor.dm [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+        return ..()
+    if(!ishuman(user)) 
+        return ..()
+
+    user.visible_message("<span class='notice'>[user] starts packing up \the [src].</span>",
+    "<span class='notice'>You start packing up \the [src]. <b>This will interrupt the process.</b></span>")
+    if(!user.get_active_held_item() && do_after(user, activation_time*1.2, TRUE, 5, BUSY_ICON_FRIENDLY,, TRUE))
+        reset_state()
+        user.put_in_hands(src)
+    . = ..()
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
    return ..()
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+
+    if(!ishuman(user)) 
+        return
+
+    if(!user.mind)
+        to_chat(user, "<span class='warning'>It doesn't seem to do anything for you.</span>")
+        return
+
+    var/area/A = get_area(user)
+    if(A && istype(A) && A.ceiling >= CEILING_METAL)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
    if(istype(A) && A.ceiling >= CEILING_METAL)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+        to_chat(user, "<span class='warning'>You have to be outside or under a glass ceiling to activate this.</span>")
+        return
+
+    user.visible_message("<span class='notice'>[user] starts setting up [src] on the ground.</span>",
+    "<span class='notice'>You start setting up [src] on the ground and inputting all the data it needs.</span>")
+
+    if(do_after(user, activation_time, TRUE, 5, BUSY_ICON_FRIENDLY,, TRUE))
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Early reteurn

File: code/game/objects/items/gamemode/survivor.dm [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+/obj/item/laptop/rescue/proc/reset_state()
+    icon_state = "tandy0"
+    w_class = initial(w_class)
+    anchored = FALSE
+    SetLuminosity(initial(luminosity))
+    update_icon()
+
+
+/obj/item/laptop/rescue/attack_alien(mob/living/carbon/Xenomorph/M)
+    if(anchored == FALSE)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
    if(!anchored)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+        to_chat(M, "<span class='warning'>It's already broken.</span>")
+        return
+        
+    M.animation_attack_on(src)
+    M.visible_message("<span class='danger'>\The [M] slices [src] apart!</span>", \
+        "<span class='danger'>You slice [src] apart!</span>", null, 5)
+    playsound(loc, "alien_claw_metal", 25, 1)
+
+    if (anchored == FALSE)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
    if (!anchored)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+        return
+
+    reset_state()
+   
+    if(prob(10))
+        new /obj/effect/decal/cleanable/blood/oil(src.loc)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
        new /obj/effect/decal/cleanable/blood/oil(loc)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+    if(current_hp <= 0)
+        to_chat(M, "<span class='warning'>It's already broken.</span>")
+        return
+
+    M.animation_attack_on(src)
+    M.visible_message("<span class='danger'>\The [M] slices [src] apart!</span>", \
+        "<span class='danger'>You slice [src] apart!</span>", null, 5)
+    playsound(loc, "alien_claw_metal", 25, 1)
+
+    if (activated == FALSE)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
    if (!activated)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+        return
+
+    current_hp -= max(0, rand(15, 30))
+    if(current_hp <= 0)
+        log_game("[key_name(M)] destroyed \the [src].")
+        reset_state()
+   
+    if(prob(10))
+        new /obj/effect/decal/cleanable/blood/oil(src.loc)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
        new /obj/effect/decal/cleanable/blood/oil(loc)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+    user.visible_message("<span class='notice'>[user] starts packing up \the [src].</span>",
+    "<span class='notice'>You start packing up \the [src]. <b>This will interrupt the process.</b></span>")
+    if(!user.get_active_held_item() && do_after(user, activation_time*1.2, TRUE, 5, BUSY_ICON_FRIENDLY,, TRUE))
+        log_game("[key_name(user)] picked up \the [src].")
+        reset_state()
+        user.put_in_hands(src)
+    return ..()
+
+
+/obj/item/beacon/rescue/attackby(obj/item/W as obj, mob/user as mob)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Make these use the proper parameters and give it proper inheritance logic
```suggestion
/obj/item/beacon/rescue/attackby(obj/item/I, mob/user, params)
```

File: code/__DEFINES/gamemode.dm (hunk 3 above)
Comment (psykzz) [reply code/__DEFINES/gamemode.dm]: My bad, will move this all.

File: code/game/objects/items/gamemode/survivor.dm (hunk 12 above)
Comment (psykzz) [reply code/game/objects/items/gamemode/survivor.dm]: It did do something but you're right it should be removed.

File: code/game/objects/items/gamemode/survivor.dm (hunk 15 above)
Comment (psykzz) [reply code/game/objects/items/gamemode/survivor.dm]: Generally i agree but that feels against the spirit of a `do_after` which make sense to have their "action" code indented.

File: code/game/objects/items/gamemode/survivor.dm (hunk 11 above)
Comment (psykzz) [reply code/game/objects/items/gamemode/survivor.dm]: Is there not other code that interacts with `obj_integritiy` I intentionally wanted to stay away from that.

File: code/game/objects/items/gamemode/survivor.dm (hunk 21 above)
Comment (psykzz) [reply code/game/objects/items/gamemode/survivor.dm]: Is the `as X` bad?

File: code/game/objects/items/gamemode/survivor.dm (hunk 15 above)
Comment (ninjanomnom) [reply code/game/objects/items/gamemode/survivor.dm]: I wouldn't consider do_after as some sort of control keyword. If it makes sense to early return with one then do so and don't be restricted by what's common.

File: code/game/objects/items/gamemode/survivor.dm (hunk 21 above)
Comment (DominikPanic) [reply code/game/objects/items/gamemode/survivor.dm]: It just does nothing but clutters the code (as only works in either input as a special construct or in verbs)

File: code/datums/emergency_calls/emergency_call.dm (hunk 4 above)
Comment (psykzz) [reply code/datums/emergency_calls/emergency_call.dm]: I handled this the wrong way because i believed this was related to the gamemode code, but ultimately this checked that using `get_random_call()` returned a valid gamemode.

I changed this instead that my gamemode code sets this variable to someone thing valid so the check passes instead.

File: code/modules/mob/new_player/new_player.dm [hunk 22]
```diff
@@ -338,6 +344,14 @@
 			var/datum/hive_status/normal/HS = GLOB.hive_datums[XENO_HIVE_NORMAL]
 			HS.stored_larva++
 
+	if(issurvivorgamemode(SSticker?.mode))
```
Comment (DominikPanic) [inline code/modules/mob/new_player/new_player.dm]: Could be made a generic handlelatespawn gamemode proc

File: code/modules/mob/new_player/new_player.dm [hunk 23]
```diff
@@ -37,10 +37,14 @@
 		output += "<p>\[ [ready? "<b>Ready</b>":"<a href='byond://?src=\ref[src];lobby_choice=ready'>Ready</a>"] | [ready? "<a href='byond://?src=[REF(src)];lobby_choice=ready'>Not Ready</a>":"<b>Not Ready</b>"] \]</p>"
 
 	else
-		output += "<a href='byond://?src=[REF(src)];lobby_choice=manifest'>View the Crew Manifest</A><br><br>"
-		output += "<p><a href='byond://?src=[REF(src)];lobby_choice=late_join'>Join the TGMC!</A></p>"
-		if(isdistress(SSticker.mode))
+		if (issurvivorgamemode(SSticker.mode))
```
Comment (DominikPanic) [inline code/modules/mob/new_player/new_player.dm]: Same, could be a gamemode proc

File: code/modules/mob/living/carbon/xenomorph/abilities.dm [hunk 24]
```diff
@@ -475,6 +475,10 @@
 
 	if(isobj(A))
 		O = A
+		if(CHECK_BITFIELD(O.resistance_flags, UNACIDABLE|INDESTRUCTIBLE))
```
Comment (DominikPanic) [inline code/modules/mob/living/carbon/xenomorph/abilities.dm]: This is not only unatomized if it's supposed to be a fix, but this is already handled through `can_use_ability`

File: code/game/objects/items/pinpointer.dm [hunk 25]
```diff
@@ -1,143 +1,114 @@
...
-				icon_state = "pinonmedium"
-			if(16 to INFINITY)
-				icon_state = "pinonfar"
-		spawn(5) .()
+    name = "Advanced Pinpointer"
+    desc = "A larger version of the normal pinpointer, this unit features a helpful quantum entanglement detection system to locate various objects that do not broadcast a locator signal."
+    var/mode = 0  // Mode 0 locates disk, mode 1 locates coordinates, 2 to find an item
+
+/obj/item/pinpointer/advpinpointer/set_target()
+    if(mode == 0)
```
Comment (DominikPanic) [inline code/game/objects/items/pinpointer.dm]: Shoiuld add defines for the mode

File: code/game/objects/items/pinpointer.dm [hunk 26]
```diff
@@ -1,143 +1,114 @@
...
+    w_class = 1
+    item_state = "electronic"
+    throw_speed = 4
+    throw_range = 20
+    matter = list("metal" = 500)
+    var/active = FALSE
+    var/target = null
+
+/obj/item/pinpointer/proc/set_target()
+    if (issurvivorgamemode(SSticker.mode))
```
Comment (DominikPanic) [inline code/game/objects/items/pinpointer.dm]: Early return

File: code/game/objects/items/gamemode/survivor.dm [hunk 27]
```diff
@@ -0,0 +1,355 @@
...
+    SetLuminosity(5)
+
+    if (issurvivorgamemode(SSticker.mode))
+        var/datum/game_mode/survivor/GM = SSticker.mode
+        GM.beacon = src
+
+    noise_timer_id = addtimer(CALLBACK(src, .proc/make_noise), 7 SECONDS, TIMER_LOOP|TIMER_STOPPABLE)
+    beacon_timer_id = addtimer(CALLBACK(src, .proc/call_distress_team), distress_timer, TIMER_UNIQUE|TIMER_STOPPABLE)
+    
+    var/mob/M
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Put it inside the loop

File: code/game/objects/items/gamemode/survivor.dm [hunk 28]
```diff
@@ -0,0 +1,355 @@
...
+                to_chat(user, "<span class='warning'>\The [C] doesn't have enough charge!</span>")
+                return
+            if (user.transferItemToLoc(W, src))
+                required_components -= R
+                internal_components += W
+                user.visible_message("<span class='notice'>[user] installed \the [C]</span>","<span class='notice'>You installed \the [C]</span>")
+                break
+
+    // repair dmg
+    if(iswelder(W) && current_hp < max_hp)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Should change those Ws into Is since you changed the name of the paramter

File: code/game/objects/items/gamemode/survivor.dm [hunk 29]
```diff
@@ -0,0 +1,355 @@
...
+    if(current_hp <= 0)
+        log_game("[key_name(M)] destroyed \the [src].")
+        reset_state()
+   
+    if(prob(10))
+        new /obj/effect/decal/cleanable/blood/oil(loc)
+    
+
+// TODO: Handle other click types (altclick etc)
+/obj/item/beacon/rescue/attack_hand(mob/user as mob)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
/obj/item/beacon/rescue/attack_hand(mob/user)
```

File: code/game/objects/items/gamemode/survivor.dm (hunk 11 above)
Comment (DominikPanic) [reply code/game/objects/items/gamemode/survivor.dm]: All objects should be handled through that

Comment (Rohesie) [reply code/game/objects/items/gamemode/survivor.dm]: The question is: does this represent object integrity or not? In this case it seems so.
It doesn't mean it needs to be destroyed once it reaches zero or lower. But if the way to lower this is through damage, then it should definitely be integrity.

**Comment by Rohesie** [comment]
The concept is great, but in the long-run we'll want roles to be assigned through regular jobs, not through special ones, so we can cycle between different game modes more smoothly.

File: code/game/objects/items/pinpointer.dm (hunk 26 above)
Comment (psykzz) [reply code/game/objects/items/pinpointer.dm]: Thats not possible for this proc, unless you can explain it a little more?

**Comment by DominikPanic** [comment]
DMM since map freeze

File: code/game/gamemodes/survivor.dm [hunk 30]
```diff
@@ -0,0 +1,460 @@
...
+
+
+/datum/game_mode/survivor/new_player_topic(mob/M, href, href_list[])
+    switch(href_list["lobby_choice"])
+        if("late_join_survivor")
+            if (world.time > (SSticker.round_start_time + late_join_time))
+                to_chat(src, "<span class='warning'>It's too late to join this round.</span>")
+                return
+
+            if (!M.mind)
```
Comment (DominikPanic) [inline code/game/gamemodes/survivor.dm]: Every mob should have a mind at all times, it may be better to rather stack_trace here

File: code/game/gamemodes/survivor.dm [hunk 31]
```diff
@@ -0,0 +1,460 @@
...
+    return TRUE
+
+
+/datum/game_mode/survivor/post_setup()
+    . = ..()
+
+    spawn_queen_in_ovi()
+    
+    // Transfer everyone to their bodies and delete existing
+    for(var/mob/new_player/player in GLOB.mob_list)
```
Comment (DominikPanic) [inline code/game/gamemodes/survivor.dm]: dead_mob_list is smaller so that may be more optimized

File: code/game/gamemodes/survivor.dm [hunk 32]
```diff
@@ -0,0 +1,460 @@
...
+            if(HIGH_POP)
+                ratio = rand(5,50)
+            if(MED_POP)
+                ratio = rand(25,50)
+            if(LOW_POP)
+                ratio = rand(50,100)
+        A.cell.charge = A.cell.maxcharge * (ratio / 100)
+    */
+
+    for (var/I in GLOB.machines)
```
Comment (DominikPanic) [inline code/game/gamemodes/survivor.dm]: these exist:
GLOBAL_LIST_EMPTY(apcs_list)
GLOBAL_LIST_EMPTY(airlocks)

File: code/game/gamemodes/survivor.dm [hunk 33]
```diff
@@ -0,0 +1,460 @@
...
+        N.new_character = H
+
+    M.transfer_to(H, TRUE)
+    H.client.prefs.copy_to(H)
+
+    var/job_title
+    var/role_guide
+    var/survivor_job = pick(subtypesof(/datum/job/survivor))
+
+    switch(rand(0, 100))
```
Comment (DominikPanic) [inline code/game/gamemodes/survivor.dm]: Why not 1 to 4?

File: code/game/gamemodes/survivor.dm [hunk 34]
```diff
@@ -0,0 +1,460 @@
...
+            role_guide = "You have extra medical supplies, that could be useful if you or someone else gets hurt..."
+        if (50 to 75)
+            survivor_job = /datum/job/survivor/security
+            job_title = SURVIVOR_FIGHTER
+            role_guide = "You have found a weapon! Hope you don't have to use it too soon..."
+        if (75 to 100)
+            job_title = SURVIVOR_SCAV
+            role_guide = "You have a radio and pinpointer! Maybe this will lead you to what you need most..."
+
+    var/datum/job/J = new survivor_job
```
Comment (DominikPanic) [inline code/game/gamemodes/survivor.dm]: Do not instance new jobs, use SSjob GetJobType

File: code/game/gamemodes/survivor.dm [hunk 35]
```diff
@@ -0,0 +1,460 @@
...
+
+    H.set_rank(J.title)
```
Comment (DominikPanic) [inline code/game/gamemodes/survivor.dm]: AssignRole is better, set_rank is meant for admin-only interactions and thus is inefficient

File: code/game/gamemodes/survivor.dm [hunk 36]
```diff
@@ -0,0 +1,460 @@
...
+            return
+        if (world.time - last_beacon_announce < 2 MINUTES)
+            return
+        annouce_beacon_location()
+
+
+/datum/game_mode/survivor/proc/annouce_beacon_location()
+    last_beacon_announce = world.time
+    var/area/beacon_area = get_area(beacon.loc)
+    var/mob/M
```
Comment (DominikPanic) [inline code/game/gamemodes/survivor.dm]: Put it in the loop

File: code/game/objects/items/gamemode/survivor.dm [hunk 37]
```diff
@@ -0,0 +1,354 @@
...
+/obj/item/laptop/rescue/update_icon()
+    icon_state = "tandy0"
+    cut_overlays()
+    if(anchored)
+        icon_state = "tandy1"
+        add_overlay("tandy1o_wifi")
+
+    return ..()
+
+/obj/item/laptop/rescue/attack_hand(mob/user as mob)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
/obj/item/laptop/rescue/attack_hand(mob/user)
```

File: code/game/objects/items/gamemode/survivor.dm [hunk 38]
```diff
@@ -0,0 +1,354 @@
...
+
+    // Timer to remove the distress beacon if we get distroyed
+    var/noise_timer_id = null
+    var/beacon_timer_id = null
+
+
+/obj/item/beacon/rescue/examine()
+    . = ..() // show parent examines (if any) first
+    if (beacon_timer_id)
+        to_chat(usr, "<span class='notice'>It shows [timeleft(beacon_timer_id) / 10] seconds left.</span>")
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Examine has a parameter, you should define it and use it (user) rather than usr.

File: code/game/objects/items/gamemode/survivor.dm [hunk 39]
```diff
@@ -0,0 +1,354 @@
...
+    if (issurvivorgamemode(SSticker.mode))
+        var/datum/game_mode/survivor/GM = SSticker.mode
+        GM.beacon = null
+    
+    if (dump_contents && length(internal_components))
+        required_components = initial(required_components)
+        for (var/I in internal_components)
+            var/turf/T = get_step(src, pick(cardinal))
+            if (T.obscured)
+                usr.transferItemToLoc(I, src)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Pass the user and use that

File: code/game/objects/items/gamemode/survivor.dm [hunk 40]
```diff
@@ -0,0 +1,354 @@
...
+            var/mob/M = i
+            SEND_SOUND(M, sound('sound/misc/notice2.ogg'))
+            to_chat(M, "<h2 class='alert'>MESSAGE RECIEVED</h2>")
+            to_chat(M, "<span class='alert'>We have lost signal with your beacon! Get it set back up or we'll never find you.</span>")
+        log_game("[src] was stopped. Timer had [distress_timer] time left.")
+
+    STOP_PROCESSING(SSobj, src)
+
+
+/obj/item/beacon/rescue/attack_alien(mob/living/carbon/Xenomorph/M)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Use X for Xenomorphs

File: code/game/objects/items/gamemode/survivor.dm [hunk 41]
```diff
@@ -0,0 +1,354 @@
...
+    return ..()
+
+
+/obj/item/beacon/rescue/attackby(obj/item/I, mob/user, params)
+    if(!anchored || activated)
+        return
+
+    // Install components
+    for (var/R in required_components)
+        if (istype(I, R))
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Early continue

File: code/game/objects/items/gamemode/survivor.dm [hunk 42]
```diff
@@ -0,0 +1,354 @@
...
+    if(iswelder(I))
+        if (obj_integrity == max_integrity)
+            to_chat(user, "That doesn't look damaged!")
+            return 
+
+        if(!do_after(user, 1 SECONDS, TRUE, 5, BUSY_ICON_BUILD,, TRUE))
+            return
+
+        var/obj/item/tool/weldingtool/WT = I
+        if(WT.remove_fuel(0, user))
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]: Early return

File: code/game/objects/items/gamemode/survivor.dm [hunk 43]
```diff
@@ -0,0 +1,354 @@
...
+    for (var/R in required_nearby)
+        var/obj/found = locate(R) in nearby
+        if (found?.anchored)
+            nearby_setup++
+
+    if(!length(required_components) && obj_integrity == max_integrity && nearby_setup == length(required_nearby))
+        log_game("[key_name(user)] activated \the [src].")
+        activate_beacon(user)
+
+/obj/item/beacon/rescue/proc/activate_beacon(mob/user as mob)
```
Comment (DominikPanic) [inline code/game/objects/items/gamemode/survivor.dm]:
```suggestion
/obj/item/beacon/rescue/proc/activate_beacon(mob/user)
```

File: code/game/gamemodes/survivor.dm (hunk 33 above)
Comment (psykzz) [reply code/game/gamemodes/survivor.dm]: Because then you wouldn't have something to review (i fucked up)

File: code/game/gamemodes/survivor.dm (hunk 30 above)
Comment (psykzz) [reply code/game/gamemodes/survivor.dm]: Added the `stack_trace`

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 44]
```diff
@@ -426,6 +426,50 @@ to_chat will check for valid clients itself already so no need to double check f
...
+
+/datum/hive_status/normal/proc/spawn_larva(mob/xeno_candidate)
+	if(!xeno_candidate?.mind)
+		return FALSE
+
+	if(!stored_larva || !istype(living_xeno_queen))
+		to_chat(xeno_candidate, "<span class='warning'>Something went awry. Can't spawn at the moment.</span>")
+		return FALSE
+
+	var/mob/living/carbon/Xenomorph/Larva/new_xeno = new /mob/living/carbon/Xenomorph/Larva(living_xeno_queen.loc)
```
Comment (psykzz) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm]: lowercase this path

**Comment by DominikPanic** [comment]
Still plan to work on this?

**Comment by psykzz** [comment]
definitely plan to, just will happen after the current work on the crash gamemode go out

**Comment by psykzz** [comment]
I understand its also massively conflicting but the conflicts will be easy to manage (as i pulled code from here for the gamemode refactor)

**Comment by psykzz** [comment]
This is pending the crash gamemode, put on hold until that is merged then ill work on this again.

**Comment by Rohesie** [comment]
I'd close the PR and re-open it when it's no longer.

**Comment by psykzz** [comment]
Closing until its time to work on again.
